src/Utility/Hash.php

```diff
@@ -1146,6 +1147,179 @@ public static function mergeDiff(array $data, array $compare): array
         return $data;
     }
 
+    /**
+     * Validates data against provided schema.
+     *
+     * @param array $data Data to validate
+     * @param array $schema Schema to validate against
+     * @return array
+     */
+    public static function validate(array $data, array $schema): array
```

Comment:
I like this feature quite a bit, and think it has potential to help us build safer applications. I wonder if this shouldn't be its own class though? I know `Hash` has many oddly shaped responsibilities but I think this is substantial enough that we could give it some more space.

Reply:
It kind of makes sense as a Hash function because we don't support lists as the top-level array. However, there are many protected methods used to implement it.

I thought about making it a separate repo even, but cakephp code will be tied to a specific schema version. 

What would we call it?

Reply:
> However, there are many protected methods used to implement it.

Yeah the number of protected methods make me think that it could be a separate class.

> What would we call it?

Some possible options I can think of:

* ShapeValidator
* SchemaValidator
* ShapeValidation

We could continue to have `Hash::validate()` or `Validation::arrayShape()` and then delegate to the separate class.